DX-61688: AES_DECRYPT function reliably segfaults the JVM

schavan6 opened the pull request:
When aes_decrypt function throws a runtime error, a null pointer was returned which caused the segmentation fault.

Fix - Removed the error-causing null pointer return, replaced with a meaningful error return.

schavan6 committed:
DX-61688: AES_DECRYPT function reliably segfaults the JVM

When aes_decrypt function throws a runtime error, a null pointer was returned which caused the segmentation fault.

Fix - Only Function error context is set in case of a runtime error, removed the error-causing null pointer return.

schavan6 committed:
DX-61688: AES_DECRYPT function reliably segfaults the JVM

When aes_decrypt function throws a runtime error, a null pointer was returned which caused the segmentation fault.

Fix - Removed the error-causing null pointer return.

schavan6 committed:
Update encrypt_utils_test.cc

Removed whitespace

vvellanki commented on cpp/src/gandiva/gdv_function_stubs.cc:
```diff
@@ -363,8 +363,8 @@ const char* gdv_fn_aes_decrypt(int64_t context, const char* data, int32_t data_l
   } catch (const std::runtime_error& e) {
-    gdv_fn_context_set_error_msg(context, e.what());
-    return nullptr;
+    std::string err_msg = "Error occurred while decrypting ciphertext.";
```
Please include the original error message as much as possible. For e.g. something like
"Error occurred while decrypting ciphertext: %s", e.what()

vvellanki replied:
Can you also search for this pattern in other functions and ensure that we fix those instances also?

Other functions that catch runtime_error might have similar issues

schavan6 committed:
DX-61688: AES_DECRYPT function reliably segfaults the JVM

When aes_decrypt function throws a runtime error, a null pointer was returned which caused the segmentation fault.

Fix - Removed the error-causing null pointer return.

Amend to the main commit - Rewrote error message and corrected the same case for aes_encrypt as well.

schavan6 commented:
@vvellanki Thank you for reviewing this. I have accordingly changed the error message and corrected only other case I found where a null pointer was returned on a runtime error.

lriggs commented on cpp/src/gandiva/gdv_function_stubs.cc:
```diff
@@ -332,8 +332,8 @@ const char* gdv_fn_aes_encrypt(int64_t context, const char* data, int32_t data_l
-    gdv_fn_context_set_error_msg(context, e.what());
-    return nullptr;
+    std::string err_msg = "Following error occurred while encrypting ciphertext - " + std::string(e.what());
+    gdv_fn_context_set_error_msg(context, err_msg.data());
```
Since this falls through to returning 'ret', is there a chance that ret would be invalid? In other words, I'm wondering what cases return a valid 'ret' pointer, but the aes_encrypt call fails?

schavan6 replied:
In this case of a runtime exception (which occurs when wrong decryption key is given), a valid 'ret' is returned, but the error set on context is correctly shown on UI. Returning null pointer instead of this, results in a segmentation fault.
